src/main/java/net/minecraftforge/client/render/BatchedTileEntityRenderer.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2018.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.client.render;
+
+import net.minecraft.client.renderer.Tessellator;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class BatchedTileEntityRenderer
+{
+
+    private static final Map<BatchedBufferConfig, Tessellator> BATCHED_BUFFERS = new LinkedHashMap<>();
+    private static boolean drawingBatch = false;
+
+    static
+    {
+        registerBatchedVertexFormat(BatchedBufferConfig.BLOCK);
+    }
+
+    public static void registerBatchedVertexFormat(BatchedBufferConfig config)
+    {
+        BATCHED_BUFFERS.put(config, new Tessellator(0x20000));
```

Comment:
can we make this idempotent? i.e. do nothing if the specified config is already registered

if we implement hash/eq as I mentioned above, and multiple people call register with equivalent configs, the existing tessellator's buffer will leak. It would also just leak if someone accidentally called this on their own more than once

Reply:
I wrestled with this problem recently writing a soon-to-be-released API core mod (https://github.com/grondag/Acuity) that allows for shaders and extended vertex formats during *block* rendering. (No TESR required at all.). I wanted performant multi-layered textures without Z-fighting or transparency sorting. I also wanted to employ shaders at scale for some fancy renders with good performance (and without forcing people to use a shaders mod or OptiFine.) It will be released along with the mods that use it after some more testing. 

Ultimately I decided to offer a just three vertex formats with the hope that one them should be "good enough" for any given use case.  The lighting.-related parts of the format may be dynamic depending on lighting model. (Right now it just emulates the vanilla lighting model, but future versions could support a better global illumination model and colored block lighting.)

Fewer vertex formats means fewer GL state changes and lets the mod do more of the work in an optimal way, like automatically handling vertex binding and state enable/disable. A single vertex buffer is used for each layer - the mod pack multiple formats  and pipelines into a single buffer and separately tracks offsets for each render pipeline.

What uses cases would require arbitrarily numerous and complex vertex formats?  Most elements can be repurposed based on need, especially if using shaders and generic attribute bindings. 

Side note: I didn't think my mod would meet the standard of broad applicability expected of a pull request, and it currently does not fully eliminate the need for TESRs - it doesn't handle animation, for example, nor will it handle renders that need per-block information.  But if the code is of any use have at it.

Reply:
@williewillus Sure, I'll add a containsKey check. But as I said, there may still be a need for different vertex buffers for the same vertex format with the same gl mode, as the IBufferDrawer might need a different GL state. If there was a simple way for mods to define there needed GL calls, this could be possible. But I didn't came up with something about this yesterday.